### PR TITLE
Exclusive Consumer Feature with connection Fail-Over 

### PR DESCRIPTION
**Pull request by IndunilRathnayake**
Added the Exclusive Consumer Feature with connection Fail-Over and Andes Queue creation by extracting informaton(enability of exclusive consumer feature) through MBeans

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 1]
```diff
@@ -69,6 +70,30 @@
 
     private ClusterSubscriptionProcessor clusterSubscriptionProcessor;
 
+    // To store active exclusive consumer for each queue
+    private Map<String, LocalSubscription> activeExclusiveConsumersForQueue = new HashMap<String, LocalSubscription>();
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Shall we explain the key and the value in the Map?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 2]
```diff
@@ -184,6 +209,44 @@ public SubscriptionStore() throws AndesException {
     }
 
     /**
+     * @param destination queue name
+     * @param isTopic     FALSE if checking queues
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: FALSE if checking queues doesn't make sense to me. What about a comment like whether a topic or a queue? WDYT?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 3]
```diff
@@ -184,6 +209,44 @@ public SubscriptionStore() throws AndesException {
...
+     * @throws AndesException
+     */
+    public Collection<LocalSubscription> getSubscriptionsForDeliver(String destination,
+                                                                    boolean isTopic)
+            throws AndesException {
+
+        if (null != activeExclusiveConsumersForQueue.get(destination) && !isTopic) {
+            LocalSubscription exclusiveSubscriber = activeExclusiveConsumersForQueue.get(destination);
+            ArrayList<LocalSubscription> subscribers = new ArrayList<LocalSubscription>();
+            subscribers.add(exclusiveSubscriber);
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Its good to have one return statement for the method.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 4]
```diff
@@ -69,6 +70,30 @@
...
     private ClusterSubscriptionProcessor clusterSubscriptionProcessor;
 
+    // To store active exclusive consumer for each queue
+    private Map<String, LocalSubscription> activeExclusiveConsumersForQueue = new HashMap<String, LocalSubscription>();
+
+    // To store all the subscribers, subscribed for exclusive consumer feature enabled queues
+    private Map<String, TreeSet<LocalSubscription>> allExclusiveSubscribers = new HashMap<String, TreeSet<LocalSubscription>>();
+
+
+    class SubscriptionComparator implements Comparator<LocalSubscription> {
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Better to have a comment here for the class.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 5]
```diff
@@ -69,6 +70,30 @@
...
+        @Override
+        public int compare(LocalSubscription sub1, LocalSubscription sub2) {
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Can have a comment as
/**
* {@inheritDoc}
*/

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 6]
```diff
@@ -609,6 +672,50 @@ public synchronized void createDisconnectOrRemoveLocalSubscription(LocalSubscrip
             channelIdMap.remove(subscription.getChannelID());
         }
 
+
+        List<AndesQueue> allQueuesStored = andesContextStore.getAllQueuesStored();
+        for (AndesQueue andesQueue : allQueuesStored) {
+            if (subscription.isDurable() && !subscription.isBoundToTopic() && andesQueue.isExclusiveConsumer && andesQueue.queueName.equalsIgnoreCase(subscription.getTargetQueue())) {
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: This line may be exceeding the margin. Wrap if it exceeds.

Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: This if condition checks lots of conditions and it reduce the readability. Can simplify this?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java [hunk 7]
```diff
@@ -184,6 +209,44 @@ public SubscriptionStore() throws AndesException {
...
+            ArrayList<LocalSubscription> subscribers = new ArrayList<LocalSubscription>();
+            subscribers.add(exclusiveSubscriber);
+            return subscribers;
+        } else {
+            return this.getActiveLocalSubscribers(destination, isTopic);
+        }
+    }
+
+    /**
+     * Remove the mapping for the queue name (exclusive enabled), when the queue get deleted and if there are no subscribers for that queue
```
Comment (hemikak) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionStore.java]: Wrap if the comment is exceeding the margin.

**Pull request closed without merging**